### ISSUE TITLE
caml_remove_global_root in finalizer spins on roots_mutex

### ISSUE THREAD

**Issue by tbrk** [ocaml/ocaml#11865]
I'm testing the Sundials/ML library against OCaml 5.0.0. One of the examples spins forever after the following sequence:
1. The GC is triggered and `caml_scan_global_young_roots` locks `roots_mutex`.
2. During the subsequent scan, one of my finalizers in C is called and it invokes `caml_remove_global_root`.
3. The resulting call to `caml_delete_global_root` requests `roots_mutex` and spins until ^C.

The smallest failing example that I have been able to produce is as follows. Sorry it's not simpler.

`block_ml.c`:
```
#include <caml/alloc.h>
#include <caml/memory.h>

typedef struct cdata { value backlink; } *pcdata;
#define PCDATA(v) (*(pcdata *)Data_custom_val(v))

void mlfinalize(value val)
{
    pcdata v = PCDATA (val);
    caml_remove_global_root(&(v->backlink));
    free(v);
}

CAMLprim value mlwrap(value payload)
{
    CAMLparam1(payload);
    CAMLlocal1(vcptr);

    pcdata v = calloc(1, sizeof(struct cdata));
    v->backlink = payload;
    caml_register_global_root(&(v->backlink));

    vcptr = caml_alloc_final(1, mlfinalize, 1, 30);
    PCDATA(vcptr) = v;

    CAMLreturn(vcptr);
}
```

`block.ml`:
```
type ptr
type 'data t = { payload: 'data; cptr: ptr }

external c_wrap : 'a -> ptr = "mlwrap"
let make v = { payload = v; cptr = c_wrap v }

type gdata = ..
type gdata += RA of int
type gdata += VV of gdata t

let rec make_ra v = make (RA v)
and make_vv v = make (VV v)
and clone { payload; _ } =
  match payload with
  | RA v -> make_ra v
  | VV nv -> make_vv (clone nv)
  | _ -> assert false

let _ =
  let y = make_ra 7 in
  let x = make_vv y in
  let _ = clone x in
  Gc.major ();
  Gc.full_major ();
  let _ = clone y in
  Gc.minor ()
```
To reproduce, compile with `ocamlopt -o block block_ml.c block.ml`, run `block`, and terminate it with Ctrl-C. The description above summarizes the back trace generated by valgrind or gdb.

It's always possible that the C interface code is not correct, but I can't see the problem...

**Comment by gasche** [comment]
This is clearly a deadlock bug in the runtime, rather than a problem with your C code.

(I recently wrote a version of the global_roots API that does not use a global lock, #11826. I believe that it would solve your issue, but it is probably not the right solution.)

It may be possible to solve this issue quickly by using a recursive/reentrant mutex. As far as I know, they are never used in the runtime currently, but is there a strong reason to avoid them? (cc @xavierleroy, @kayceesrk who would know)

Another approach would be to delay the calls to finalizers to after the root scanning (not unlike how Statmemprof delays calls to its callbacks when invoked from C code). This is probably a better fix, but also rather more invasive.

**Comment by gasche** [comment]
In your report I was confused as to how it could be that a finalizer is called by `caml_scan_global_young_roots`. This function merely oldifies young global roots, it does not "delete" any root in a way that would call a finalizer. The answer is in the backtrace of your nice repro case:
-  `oldify_one` calls `alloc_shared` to get major heap space to move young roots
- `alloc_shared` calls `pool_allocate` which calls `pool_find` to locate a pool with free slots
- none of the current pools have free slots left, so `pool_find` calls `pool_sweep` to try to make some space
- this in turn frees dead values and eagerly call finalizers

I would not be surprised if there were other scenarios (than program with global roots) where calling finalizers in the middle of `oldify_one` could turn out to be an issue.

**Comment by xavierleroy** [comment]
> It may be possible to solve this issue quickly by using a recursive/reentrant mutex. As far as I know, they are never used in the runtime currently, but is there a strong reason to avoid them? 

I think recursive mutexes have their uses.  A  typical use is to protect I/O channels, so that you can lock a channel explicitly across several I/O operations, yet keep each operation protected by the lock.  Maybe `roots_mutex` is a good use of a recursive mutex, too.  

Also, note that recursive mutexes are the default mutual exclusion mechanism in Java and in the Win32 API.

On the other hand, some multithreading experts have strong opinions against recursive mutexes.  See e.g. Dave Butenhof's arguments: http://www.zaval.org/resources/library/butenhof1.html .

**Comment by bluddy** [comment]
> On the other hand, some multithreading experts have strong opinions against recursive mutexes. See e.g. Dave Butenhof's arguments: http://www.zaval.org/resources/library/butenhof1.html .

I read this, and I don't think he makes a convincing case. He's basically saying, "recursive mutexes are bad because it's just a safety valve and you're not aware of what all your code paths are doing and you could be inefficient". I think that's a strong argument *for* recursive mutexes. It is *impossible* to be aware of all code paths. We simply don't have the ability to think them all through, nor do we have the tools to do it for us. And in the few cases where we *can* handle it now, we *won't* as the application scales. It makes no sense to me to insist on a human understanding every subtlety of code interaction out of some notion of a sort of "purity". It's an argument that simply cannot scale.

**Comment by gasche** [comment]
After thinking more about this, I think that recursive mutexes are not the solution here in any case: with recursive mutexes this code example would try to insert/modify elements in the globroots skiplists in the middle of their traversal for root scanning, and this sort of reentrancy issues would also lead to incorrect behaviors -- a root registered by a finalizer could get silently dropped from the global skiplists, for example.

**Comment by gasche** [comment]
I propose a more general fix in #11868, but this is an invasive change that will require some time.

**Comment by tbrk** [comment]
Thanks @gasche for looking at this issue so quickly. From my side, there's no rush as I can work around the original problem by adding a preemptive `Gc.compact` in my test script. I'm following the technical discussions with great interest, but regrettably I don't have enough expertise to help directly. Please let me know, though, if you want me to run the Sundials/ML tests against an eventual patch (there are quite a few and the Makefile is setup to allow forcing the Gc and running with valgrind; @jun0 and I found a lot of problems this way, mostly in our own code).

**Comment by sadiqj** [comment]
One less invasive way to avoid this is to stop sweeping pools during `pool_find` in minor collections, it's not needed for correctness (we'll do it eventually anyway).

We should benchmark the impact, it could increase peak major heap usage and reduce minor heap pause times.

Option two is we snapshot the global roots while holding the mutex, drop the mutex and iterate over the snapshot.

**Comment by gasche** [comment]
This is what I'm trying to do in #11868, but it seems tricky.

**Comment by gadmm** [comment]
@tbrk As mentioned in https://github.com/ocaml/ocaml/pull/11868#issuecomment-1372521766, a decent workaround is to change the custom finalizer into an OCaml finalizer (i.e. do not register it using `caml_alloc_final` but instead by calling `Gc.finalize` somehow).

**Comment by gadmm** [comment]
As a follow-up to the discussion at #11868, I advocate to fix the remove_global_root functions to be callable from custom finalizers (i.e. not "require access to the runtime" in ownership parlance).

Here is what I had in mind using an error-checking mutex and a queue:
* delete root:
  ```c
  int success = lock();
  enqueue(root);
  if (success) clear_queue_and_unlock();
  ```
* scan roots:
  ```c
  lock();
  scan_lists(action);
  clear_queue_and_unlock();
  ```
* `lock`: using an error-checking mutex, try to lock the mutex and return 0 if we already have it
* `enqueue`: store the root to delete somehow somewhere. As it turns out, globroots.c does not use the `data` field of the skiplists, so that's free real-estate to store a linked list!
* `clear_queue_and_unlock`: actually call skiplist_remove on the elements of the queue, then release the mutex

**Comment by gasche** [comment]
I had a simpler approach in mind, which was to  copy the the roots into a local malloced array,  and thus scan them outside the critical section. Your approach is probably more efficient (especially if people use non-generational global roots), but also more work and more code, and in fact malloc probably performs very well in practice.

**Comment by gadmm** [comment]
This is what @sadiqj proposed and of course this also works. In fact this is probably enough.

**Comment by gadmm** [comment]
To be clear, are you saying that you are working on it, or do you want someone else to pick it up?

**Comment by gasche** [comment]
I am not currently working on it -- I switched to other things after closing #11868.

**Comment by gadmm** [comment]
I just had a stab at it. I think that the "copy roots and release mutex" approach does not work. Indeed, once the user removes the root, they are free to do anything they want with the initial memory location. For example, with the pattern of `malloc(sizeof(value))` followed by `register_global_root`, the release does `remove_global_roots` followed by `free`. Thus you have a race whereby `free` might have re-used the memory location by the time runtime scans the remembered root.

I have a race leading to a segfault in my implementation during tests that could be explained by this (though I did not put the finger on what causes it exactly in my runs). The previous example then convinced me that it cannot work with this approach.

**Comment by gadmm** [comment]
As for my proposition, you would think it is more robust against the race I described since it does not release the lock early. However, it is plausible that the custom finaliser could remove a global root and immediately `free` the associated memory, so the same issue presents itself. (As for making sure to delete the root as soon as possible includign from the iteration sequence, I do not see a reason that the root being removed must be different from the one on which the scanning action is being called.)

I fear one has to go back to the idea of preventing custom finalizers from running during scanning actions, if one wants to fix this (e.g. #11868, or treating custom finalizers like OCaml finalizers). I'm listening for other ideas.

**Comment by gasche** [comment]
Well, one idea is #11826 which uses concurrency-safe skiplists, and is therefore able to deal gracefully with removal in the middle of iteration. I closed the PR because the constant factor overhead was large, but it was in a setting with very heavy usage of roots not representative of most OCaml programs, it may be that the performance overhead is much less noticeable with the typical (smaller) number of roots.

**Comment by gadmm** [comment]
If I am not mistaken, it is still susceptible to the problem of unregistering the very root that is being scanned, isn't it?

**Comment by gasche** [comment]
Ah, yes, I see. I don't have other ideas to propose.

My impression is that it's not clearly worth it to fix the issue at this point:
- we have run out of "easy" fixes and the remaining ideas are more invasive
- it is possible for users to avoid it by using OCaml finalisers directly
  (but: I would be curious to hear of performance result in a real case to know if it makes a difference)
- in the medium term, boxroots would be available for people that prefer to stick with C finalisers

**Comment by gadmm** [comment]
I think the main issue is one of backwards-compatibility of a useful and used pattern that was not clearly illicit until now, as several points in the discussion #11868 lead to conclude.

Reimplementing custom finalizers with a queue similar to OCaml finalizers (i.e. non-finalized custom blocks do not end up sweepable in the first place) seems the most promising at this point, I do not know how hard it is, and it possibly comes with other problems. As you say.

**Comment by gadmm** [comment]
How about running custom finalizers in a different thread? (edit: never mind)

**Comment by gadmm** [comment]
People also use `caml_stat_free` inside finalizers (e.g. zlib). This is safe except in "pooled mode" (`OCAMLRUNPARAM=c`). Here I think the best course of action is to deprecate/disable the pooled mode (at least for user-facing functions), and guarantee that `caml_stat_free` does not "access the runtime" (it is an alias for `free`).

**Comment by gasche** [comment]
@tbrk have you, by chance, had to time to try using the proposed workaround (use `Gc.finalize` instead) in your real code base, and can you report on the performance impact, if any?

**Comment by tbrk** [comment]
@gasche : I didn't do it yet, but I will soon. I'll try to give figures for both 4.14.1 and 5.0.0. The latter will be a bit tricky because more of the examples encounter the mutex problem when run iteratively by the performance scripts.

**Comment by gadmm** [comment]
Another reason to solve it natively is that the workaround with Gc.finalize is not perfect, notably if you start taking asynchronous exceptions into account; this aspect should be easier to deal with on the C side.

I tried to implement the following strategy:
- Change `struct final` in `finalise.h` to accept either a C or an OCaml function;
- Change `alloc_custom_gen` in `custom.h` to use the finalization mechanism from `finalise.h` instead of the custom finalizers (at least for custom ops that have not been vetted to be safe for the OCaml 5 behaviour, that is, the behaviour can remain unchanged for stdlib custom blocks).

There are subtleties to deal with once you get into the code. One roadblock is how to deal with exceptions; the finaliser mechanism expects exceptions returned in the _exn format, this would not be a problem given that those were verboten before and thus `caml_raise` is not expected to appear in legacy code; except that custom ops give finalisers a return type `void` (instead of `value` as expected for an encoded exception). The other is how to best remember that a custom block is finalised with one mechanism or another.

(edit: since it was not as straightforward as I hoped, I am no longer working on it at the moment)

**Comment by gadmm** [comment]
One data point against moving C-side finalisers to behave like OCaml-side finalisers is that OCaml always runs C-side finalizers at shutdown, but OCaml-side finalisers are not guaranteed to be run before shutdown.

That is, before OCaml 5.0. For some reason the calling of C-side finalizers during shutdown was never implemented in OCaml multicore and the code path was simply deleted during the merge. This sounds like a regression. (#11909)

edit: only in "cleanup at exit" mode.

**Comment by gadmm** [comment]
Following a discussion with @gasche, a strategy could be: start from #11868 again, in order to treat specially the sweeping when requested from an unsafe location, to avoid running finalizers at that point in time. Instead, revive the custom cell. To avoid delaying its collection to the next cycle (and risk corner-cases where it takes too long to be disposed of) one could push it to the other finalizer queue at that point. The goal should simply be to make `remove_*_root` safe (not needing the capability), without other specification relaxation.

This amounts to changing the GC in order to fix `remove_*_root`, but after many attempts there is no other solution left. I am also sensitive to @gasche's point that running finalizers at those locations might have other unknown consequences.

**Comment by stedolan** [comment]
Fixed by #13584 (but feel free to reopen if this isn't enough to fix Sundials/ML)

**Comment by tbrk** [comment]
Thank you @stedolan (and everyone else for the time spent on this problem), this is great! I much prefer the idea of not having to split my finalizers across C and OCaml. They're difficult enough to think about as it stands.

I finally had a bit of time to get back into Sundials/ML today. Under OCaml 5.3.0, 1 (of 126) example deadlocks. With a recent version of the OCaml 5.4 branch (0fc7906989bb3edfff9102061860e6a0cfea8016), 0 examples deadlock. For this kind of problem, this is obviously not at all conclusive, but it is positive.
